iroh/src/downloader/test/invariants.rs

```diff
@@ -0,0 +1,83 @@
+//! Invariants for the service.
+
+#![cfg(test)]
+
+use super::*;
+
+/// invariants for the service.
+impl<G: Getter<Connection = D::Connection>, D: Dialer> Service<G, D> {
+    /// Checks the various invariants the service must maintain
+    #[track_caller]
+    pub(in crate::downloader) fn check_invariants(&self) {
+        self.check_active_request_count();
+        self.check_scheduled_requests_consistency();
+        self.check_idle_peer_consistency();
+        self.chech_concurrency_limits();
+    }
+
+    /// Checks concurrency limits are maintained.
+    #[track_caller]
+    fn chech_concurrency_limits(&self) {
+        let ConcurrencyLimits {
+            max_concurrent_requests,
+            max_concurrent_requests_per_peer,
+            max_open_connections,
+        } = &self.concurrency_limits;
+
+        // check the total number of active requests to ensure it stays within the limit
+        assert!(self.in_progress_downloads.len() <= *max_concurrent_requests);
```

Comment:
maybe add some string comments on the `assert!`s, otherwise sometimes pretty hard to parse the failures

Reply:
done in https://github.com/n0-computer/iroh/commit/aba53f2e60dd5d79f7588a9ea152237b847e0548